Mathlib/Algebra/GroupWithZero/Action/Basic.lean

```diff
@@ -0,0 +1,186 @@
+
+The hierarchy is extended further by `Module`, defined elsewhere.
+
+Also provided are typeclasses for faithful and transitive actions, and typeclasses regarding the
+interaction of different group actions,
+
+* `SMulCommClass M N α` and its additive version `VAddCommClass M N α`;
+* `IsScalarTower M N α` and its additive version `VAddAssocClass M N α`;
+* `IsCentralScalar M α` and its additive version `IsCentralVAdd M N α`.
+
+## Notation
+
+- `a • b` is used as notation for `SMul.smul a b`.
+- `a +ᵥ b` is used as notation for `VAdd.vadd a b`.
+
+## Implementation details
+
+This file should avoid depending on other parts of `GroupTheory`, to avoid import cycles.
+More sophisticated lemmas belong in `GroupTheory.GroupAction`.
+
+## Tags
+
+group action
+-/
+
+-- TODO:
+-- assert_not_exists Ring
+
+open Function
+
+variable {R R' M M' N G A B α β : Type*}
+
+section GroupWithZero
+variable [GroupWithZero α] [MulAction α β] {a : α}
+
+protected lemma MulAction.bijective₀ (ha : a ≠ 0) : Bijective (a • · : β → β) :=
+  MulAction.bijective <| Units.mk0 a ha
+
+protected lemma MulAction.injective₀ (ha : a ≠ 0) : Injective (a • · : β → β) :=
+  (MulAction.bijective₀ ha).injective
+
+protected lemma MulAction.surjective₀ (ha : a ≠ 0) : Surjective (a • · : β → β) :=
+  (MulAction.bijective₀ ha).surjective
+
+end GroupWithZero
+
+section DistribMulAction
+variable [Group G] [Monoid M] [AddMonoid A] [DistribMulAction M A]
+
+-- Porting note: this probably is no longer relevant.
+/-! Since Lean 3 does not have definitional eta for structures, we have to make sure
+that the definition of `DistribMulAction.toDistribSMul` was done correctly,
+and the two paths from `DistribMulAction` to `SMul` are indeed definitionally equal. -/
+example :
+    (DistribMulAction.toMulAction.toSMul : SMul M A) = DistribMulAction.toDistribSMul.toSMul := rfl
+
+variable (A)
+
+/-- Each element of the group defines an additive monoid isomorphism.
+
+This is a stronger version of `MulAction.toPerm`. -/
+@[simps (config := { simpRhs := true })]
+def DistribMulAction.toAddEquiv [DistribMulAction G A] (x : G) : A ≃+ A where
+  __ := toAddMonoidHom A x
+  __ := MulAction.toPermHom G A x
+
+variable (G M)
+
+/-- Each element of the group defines an additive monoid isomorphism.
+
+This is a stronger version of `MulAction.toPermHom`. -/
+@[simps]
+def DistribMulAction.toAddAut [DistribMulAction G A] : G →* AddAut A where
+  toFun := toAddEquiv _
+  map_one' := AddEquiv.ext (one_smul _)
+  map_mul' _ _ := AddEquiv.ext (mul_smul _ _)
+
+end DistribMulAction
+
+/-- Scalar multiplication as a monoid homomorphism with zero. -/
+@[simps]
+def smulMonoidWithZeroHom {α β : Type*} [MonoidWithZero α] [MulZeroOneClass β]
+    [MulActionWithZero α β] [IsScalarTower α β β] [SMulCommClass α β β] : α × β →*₀ β :=
+  { smulMonoidHom with map_zero' := smul_zero _ }
+
+section MulDistribMulAction
+
+variable [Group α] [Monoid β] [MulDistribMulAction α β]
```

Comment:
Can you file a follow-up PR to rename these variables, say \alpha to a Group and \beta to M?

Reply:
I have renamed the variables in this PR